mcruzdev/1

mcruzdev opened the pull request:
This pull request was born from #1 issue

**Adding dependencies** - https://github.com/salaboy/fmtok8s-agenda/commit/62695d2d35351676734b7592ef5177687807a6aa

I did add 2 dependencies:
- org.springframework.boot:spring-boot-starter-data-mongodb-reactive:
- de.flapdoodle.embed:de.flapdoodle.embed.mongo

**Extracting classes from AgendaService.java** - https://github.com/salaboy/fmtok8s-agenda/commit/bbbed5d6e582e4586a742d4dfe549c0d095c4789

I refactored separing resposabilities, i extracted the resource Info to InfoController, the AgendaService's main method to AgendaServiceApplication.class and create a AgendaController to solve all http requests.

**Creating reactive mongo repository** - https://github.com/salaboy/fmtok8s-agenda/commit/dd8ab93034d9ff67ff99e672f2e1cfc56cce01c6

I just created a class with a data access responsability

**Replacing in-memory repository (java.util.Set) to in-memory mongodb** - https://github.com/salaboy/fmtok8s-agenda/commit/8cbf43df63b4825ae5f7f7d33895515ca8791cf5

Was replaced the Set object to use AgendaItemRepository

**Implementing test** - https://github.com/salaboy/fmtok8s-agenda/commit/f50a261f3482415521ee51da925986f2723fa3b2

Was implemented integration tests to test the AgendaController.class

**Adding after method to drop collection AgendaItem after tests** - https://github.com/salaboy/fmtok8s-agenda/pull/2/commits/a976494c172a9a89dca5aa86815a49f30aca9512

Grow uping the test code

Commit:
Adding dependencies

Commit:
Extracting classes from AgendaService.java

Commit:
Creating reactive mongo repository

Commit:
Replacing in-memory repository ( java.util.Set ) to in-memory mongodb

Commit:
Implementing tests

Commit:
Adding after method to drop collection AgendaItem after tests

Commit:
Adjusting code after testing locally

salaboy commented on src/main/java/com/salaboy/conferences/agenda/controller/AgendaController.java:
```diff
@@ -0,0 +1,86 @@
+            throw new IllegalStateException("Something went wrong with adding the Agenda Item: " + agendaItem);
+        }
+
+        Mono<String> result = agendaItemRepository.save(agendaItem)
```
@mcruzdev do you really need to create `result` here? You can just add subscribe() to the end and return the value, without the need of creating the variable `result` right?

mcruzdev replied:
@salaboy If I add subscribe() to the end I will receive a Disposable.
I need to subscribe() because the repository doesn't save the AgendaItem

The server is not subscribing and because this is not saving the AgendaItem

mcruzdev replied:
@salaboy I solve this problem, you are correct ;)

salaboy commented on src/main/java/com/salaboy/conferences/agenda/controller/AgendaController.java:
```diff
@@ -0,0 +1,86 @@
+    @ZeebeWorker(name = "agenda-worker", type = "agenda-publish")
+    public void newAgendaItemJob(final JobClient client, final ActivatedJob job) {
+        Proposal proposal = objectMapper.convertValue(job.getVariablesAsMap().get("proposal"), Proposal.class);
+        String[] days = {"Monday", "Tuesday"};
```
@mcruzdev what do you think about extracting the time and day calculator to a separate method or a class that we can test independently?

mcruzdev replied:
It is a good idea, thanks ;)

salaboy commented on src/main/java/com/salaboy/conferences/agenda/controller/AgendaController.java:
```diff
@@ -0,0 +1,86 @@
+        int day = random.nextInt(2);
+        int time = random.nextInt(8);
+        newAgendaItem(new AgendaItem(proposal.getTitle(), proposal.getAuthor(), days[day], times[time]));
+        client.newCompleteCommand(job.getKey()).send();
```
@mcruzdev if we want to test this line here.. we need to mock it in the same way that we are mocking the ZeebeLifecycleClient in the c4p service right?

salaboy reviewed (commented):
@mcruzdev I've added some comments.. and I think that we need to mock the JobClient in the same way that we mocked the ZeebeLifecycleClient in the c4p service.. was that the screenshot that you shared in twitter?

mcruzdev commented:
@salaboy 
The image that I sent to you on twitter is about C4P's ContractVerifier tests and I already create an issue and PR solving the problem

https://github.com/salaboy/fmtok8s-c4p/pull/6

Commit:
Extracting the time and day generators and zeebeWorkerHandler

Commit:
Adding 'TestConfiguration.class' to ContractVerifierBase.java

Commit:
Creating an AgendaItemService to create an AgendaItem

Commit:
Creating a 'util' package that helps us to code a test

mcruzdev commented:
**Extracting the time and day generators and zeebeWorkerHandler**
e28f414

I extracted the generation of time and day to another class and I tested it

**Adding 'TestConfiguration.class' to ContractVerifierBase.java**
0b2bb44

In this commit, I added all mocks from TestConfiguration to ContractVerifierBase

**Creating an AgendaItemService to create an AgendaItem**
c09bc20

I created the AgendaItemService that is used by AgendaItemController.java and ZeebeWorkers.java

**Creating a 'util' package that helps us to code a test**
9da1bba

In this commit, I created a simple package that contains a simple class that creates AgendaItem objects to tests

Commit:
Changing integration test's name

salaboy commented:
@mcruzdev awesome stuff! merging now.. and testing in kubernetes !
